populator-machinery/controller.go

```diff
@@ -108,11 +113,31 @@ type controller struct {
 	recorder             record.EventRecorder
 	referenceGrantLister referenceGrantv1beta1.ReferenceGrantLister
 	referenceGrantSynced cache.InformerSynced
+	usePod               bool
+	populate             func(*PopulatorParams) (bool, error)
+	populateComplete     func(*PopulatorParams) (bool, error)
+}
+
+type PopulatorParams struct {
+	KubeClient   kubernetes.Interface
+	StorageClass *storagev1.StorageClass
+	PvcPrime     *corev1.PersistentVolumeClaim
+	Pvc          *corev1.PersistentVolumeClaim
 }
 
 func RunController(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix string,
 	gk schema.GroupKind, gvr schema.GroupVersionResource, mountPath, devicePath string,
 	populatorArgs func(bool, *unstructured.Unstructured) ([]string, error),
+) {
+	RunControllerV2(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix, gk, gvr, mountPath, devicePath, populatorArgs,
+		true, nil, nil)
+}
+
+func RunControllerV2(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath, namespace, prefix string,
```

Comment:
I think it is possible for a controller to want to support both methods? ie for some parameters, use a custom function, for other parameters, use a Pod to populate.

Reply:
+1 we will have such a req for the same controller to initiate both pod and no pod mode

Reply:
The current req that we know of are that we have a cloud provider specific implementation vs the vanilla implementation (which happens to use a pod).

Let's say for GCP right now for storage type 'foo', we do not want to spin up pods. Later for storage type 'bar' we would still need a cloud provider specific implementation but that may need pod (or deployments or anything else).  In the provider specific implementation we can infer from the storage class the storage type and take the correct route. 


I propose the following:
1. we still let the user of the lib select the vanilla or the provider specific route (and not both). We can revisit if we have such a req.
2. remove the usePod bool. use the populate arg as a signal to skip the pod creation code path. The populate here would indicate the provider specific implementation

Reply:
If we don't use a flag then the vanilla and provider specific route will be an if-else choice which won't be able to use at the same time later? Use a flag makes it easier to isolate these two options?